## Coursera Applied Data Science Capstone 
### Week 3 Assignment: part I II and III
#### Segmenting and Clustering Neighborhoods in Toronto

## Part 1

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 

So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

## Web Scraping
Here I use pandas read method to scrape the webpage.

In [2]:
import pandas as pd
import numpy as np

In [3]:
!pip install BeautifulSoup4

In [4]:
import html5lib

In [13]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df=pd.read_html(url, flavor='html5lib')[0]

In [14]:
df

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
283      M8Z         Etobicoke              Mimico NW
284      M8Z         Etobicoke     The Queensway West
285      M8Z         Etobicoke  Royal York South West
286      M8Z         Etobicoke         South of Bloor
287      M9Z      Not assigned           Not assigned

[288 rows x 3 columns]

In [15]:
df.shape

(288, 3)

The dataframe consists of 3 columns: PostalCode, Borough, and Neighborhood
It has 288 entries.

## Data Cleaning

I Only process the cells that have an assigned borough. i.e. Ignore cells with a borough that is Not assigned.

In [16]:
df=df[df.Borough != 'Not assigned']
df

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M5A  Downtown Toronto               Regent Park
6        M6A        North York          Lawrence Heights
..       ...               ...                       ...
282      M8Z         Etobicoke  Kingsway Park South West
283      M8Z         Etobicoke                 Mimico NW
284      M8Z         Etobicoke        The Queensway West
285      M8Z         Etobicoke     Royal York South West
286      M8Z         Etobicoke            South of Bloor

[211 rows x 3 columns]

77 entries were excluded under this criteria.


More than one neighborhood can exist in one postal code area. 
For example, in the table on the Wikipedia page,  M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows need to be combined into one row with the neighborhoods separated with a comma.

Here I group each postcode by 'Borough' and  get the corresponding neighbourhoods as a list by using apply(list).

In [17]:
df=df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(list)
df

Postcode  Borough    
M1B       Scarborough                                     [Rouge, Malvern]
M1C       Scarborough             [Highland Creek, Rouge Hill, Port Union]
M1E       Scarborough                  [Guildwood, Morningside, West Hill]
M1G       Scarborough                                             [Woburn]
M1H       Scarborough                                          [Cedarbrae]
                                               ...                        
M9N       York                                                    [Weston]
M9P       Etobicoke                                            [Westmount]
M9R       Etobicoke      [Kingsview Village, Martin Grove Gardens, Rich...
M9V       Etobicoke      [Albion Gardens, Beaumond Heights, Humbergate,...
M9W       Etobicoke                                            [Northwest]
Name: Neighbourhood, Length: 103, dtype: object

Join neighbourhoods in the same borough and included in a dataframe and reset the index.

In [18]:
df=df.apply(lambda x: ','.join(x)).to_frame().reset_index()
df

Postcode      Borough                                      Neighbourhood
0        M1B  Scarborough                                      Rouge,Malvern
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough                    Guildwood,Morningside,West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
..       ...          ...                                                ...
98       M9N         York                                             Weston
99       M9P    Etobicoke                                          Westmount
100      M9R    Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...
102      M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]

If a cell has a borough but No assigned neighborhood, then the neighborhood will be the same as the borough.

In [19]:
for index,row in df.iterrows():
    if row['Neighbourhood']=='Not assigned':
        row['Neighbourhood']=row['Borough']

#### here is the Final output for part I

In [20]:
df.head(20)

Postcode      Borough                                      Neighbourhood
0       M1B  Scarborough                                      Rouge,Malvern
1       M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2       M1E  Scarborough                    Guildwood,Morningside,West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7       M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8       M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9       M1N  Scarborough                         Birch Cliff,Cliffside West
10      M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11      M1R  Scarborough                                   Maryvale,Wexford
12      M1S  Scarborough                                          Agincourt
13      M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
14      M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...
15      M1W  Scarborough                                    L'Amoreaux West
16      M1X  Scarborough                                        Upper Rouge
17      M2H   North York                                  Hillcrest Village
18      M2J   North York                         Fairview,Henry Farm,Oriole
19      M2K   North York                                    Bayview Village

In [21]:
print('Data frame size =', df.shape)

Data frame size = (103, 3)
